File: Algo/Binary_Search/BS.cpp
#include <vector>
using namespace std;

// The element in arr is unique.
int bs_1(vector<int> & arr, int target) {
    int left = 0, right = arr.size();
    while (left < right) {
        int mid = (right - left) / 2 + left;
        if (arr[mid] == target)
            return mid;
        if (arr[mid] < target)
            left = mid + 1;
        else
            right = mid;
    }
    return -1;
}

// not unique
int bs_2(vector<int> & arr, int target) {
    int left = 0, right = arr.size();
    while (left < right) {
        int mid = (right - left) / 2 + left;
        if (arr[mid] <= target)
            left = mid + 1;
        else
            right = mid;
    }
    return arr[left] == target? left: -1;
}

int main() {
    return 0;
}

File: Binary_Search/BS.py


# values in arr are unique
# return index if found else -1
def Binary_Search_1(arr: list, target: int):
    left, right = 0, len(arr)

    while left < right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        if arr[mid] < target:
            left = mid + 1
        else:
            right = mid
    return -1

# There are duplicate values in arr
# Find the most left one
def Binary_search_2(arr: list, target: int):
    left, right = 0, len(arr)
    while left < right:
        mid = (left + right) // 2
        if arr[mid] < target:
            left = mid + 1
        else:
            right = mid
    
    return -1 if arr[left] != target else left